src/morpho-blue/VirtualHealthFactor.sol

```diff
@@ -0,0 +1,104 @@
+    /// @param user The address of the user whose health factor is being calculated.
+    /// @return healthFactor The calculated health factor.
+    function userHealthFactor(MarketParams memory marketParams, Id id, address user)
+        public
+        view
+        returns (uint256 healthFactor)
+    {
+        uint256 collateralPrice = IOracle(marketParams.oracle).price();
+        uint256 collateral = morpho.collateral(id, user);
+        uint256 borrowed = morpho.expectedBorrowAssets(marketParams, user);
+
+        uint256 maxBorrow = collateral.mulDivDown(collateralPrice, ORACLE_PRICE_SCALE).wMulDown(marketParams.lltv);
+
+        if (borrowed == 0) return type(uint256).max;
+        healthFactor = maxBorrow.wDivDown(borrowed);
+    }
+
+    /// @notice Calculates the health factor of a user after a virtual repayment.
+    /// @param marketParams The parameters of the market.
+    /// @param id The identifier of the market.
+    /// @param user The address of the user whose health factor is being calculated.
+    /// @param repaymentAmount The amount of assets to be virtually repaid.
+    /// @return healthFactor The calculated health factor after the virtual repayment.
+    function userHealthFactorAfterVirtualRepayment(
+        MarketParams memory marketParams,
+        Id id,
+        address user,
+        uint256 repaymentAmount
+    ) public view returns (uint256 healthFactor) {
+        uint256 collateralPrice = IOracle(marketParams.oracle).price();
+        uint256 collateral = morpho.collateral(id, user);
+        uint256 borrowed = morpho.expectedBorrowAssets(marketParams, user);
+
+        // Revert if repaymentAmount exceeds the borrowed amount
+        require(repaymentAmount <= borrowed, "Repayment amount exceeds borrowed amount");
+
+        uint256 newBorrowed = borrowed - repaymentAmount;
+        uint256 maxBorrow = collateral.mulDivDown(collateralPrice, ORACLE_PRICE_SCALE).wMulDown(marketParams.lltv);
+
+        return newBorrowed == 0 ? type(uint256).max : maxBorrow.wDivDown(newBorrowed);
+    }
+
+    /// @notice Calculates the health factor of a user after a virtual borrow.
+    /// @param marketParams The parameters of the market.
+    /// @param id The identifier of the market.
+    /// @param user The address of the user whose health factor is being calculated.
+    /// @param borrowAmount The amount of assets to be virtually borrowed.
+    /// @return healthFactor The calculated health factor after the virtual borrow.
+    function userHealthFactorAfterVirtualBorrow(
+        MarketParams memory marketParams,
+        Id id,
+        address user,
+        uint256 borrowAmount
+    ) public view returns (uint256 healthFactor) {
+        uint256 collateralPrice = IOracle(marketParams.oracle).price();
+        uint256 collateral = morpho.collateral(id, user);
+        uint256 borrowed = morpho.expectedBorrowAssets(marketParams, user);
+
+        uint256 newBorrowed = borrowed + borrowAmount;
```

Comment:
Note here you can overflow for a similar reason than during repay. Yet we don't revert with a dedicated message in this case (inconsistency). I would prefer not reverting during repay